Friend request button [PAR-78]

Hashiphoto opened the pull request:
Placed the button on the library page and friend library page. The button changes appearance and function depending on what that person's status is relative to the current user. 

Edit: A couple of things to note. I renamed the buttons because they were confusingly similar. The FriendResponseButtons are the "Accept"/"Decline" buttons for `waiting` friend request status. The FriendStatusButton adapts to be the necessary button (Request friendship, respond to request, friends already, waiting for response).

Hashiphoto committed:
FriendRequestBtns more versatile [PAR-78]

Hashiphoto committed:
Merge branch 'master' into friend-request-button

Hashiphoto committed:
Placed FriendStatusButton on both library pages [PAR-78]

Hashiphoto commented:
Just realized it inconsistently breaks. I'll request review again later once I've fixed it

Hashiphoto committed:
Pad button, user obj in friend list [PAR-78]

Hashiphoto commented:
Issue has been fixed. I'll post some screenshots in Teams of what the UI looks like

Hashiphoto committed:
Button updates on Accept/Reject [PAR-78]

lemniscate8 commented on src/FriendshipResponseButtons.tsx:
```diff
@@ -0,0 +1,98 @@
+    throw new Error("Row lacks valid data");
+  }
+
+  useEffect(() => {}, []);
```
Leftover `useEffect`

lemniscate8 commented on src/FriendshipStatusButton.tsx:
```diff
@@ -0,0 +1,84 @@
+        </Button>
+      )}
+      {friend.status === "requested" && (
+        <Button variant="info" disabled onClick={requestFriendship}>
```
If this button is always disabled it doesn't need an onClick property.

lemniscate8 commented on src/FriendshipStatusButton.tsx:
```diff
@@ -0,0 +1,84 @@
+  }
+
+  return (
+    <PaddedDiv>
```
If only one of these buttons can appear on the page, maybe use a switch statement or a sequence of if-else statements.

lemniscate8 commented on src/FriendshipStatusButton.tsx:
```diff
@@ -0,0 +1,84 @@
+import { User } from "./ourtypes";
+
+interface FRBProps {
+  rowitem?: User;
```
I only used these props because I thought this button might be used in the autotable, but if it isn't, you could rename this property to something more specific and make it required (without the ?). The checks in line 23 an a lot of other places wouldn't be necessary.

lemniscate8 commented on src/FriendshipStatusButton.tsx:
```diff
@@ -0,0 +1,84 @@
+  padding-bottom: 30px;
+`;
+
+const FriendshipRequestButton: React.FC<FRBProps> = ({
```
Maybe rename to FriendshipStatusButton?

lemniscate8 commented on src/FriendshipResponseButtons.tsx:
```diff
@@ -0,0 +1,98 @@
+  });
+}
+
+const FriendRequestButtons: React.FC<ButtonGroupProps> = ({
```
Maybe this isn't the place for this (we can always refactor later), but this component could probably be split into two independent buttons.

lemniscate8 commented on src/FriendshipStatusButton.tsx:
```diff
@@ -0,0 +1,84 @@
+    if (!friend) {
+      return;
+    }
+    onClick({ ...friend, status: "requested" });
```
Should this be set even if the fetch fails? The pattern we've been using usually puts this sort of behavior inside a `.then` call.

lemniscate8 reviewed (changes requested):
Just had some small suggestions. I can see what you meant in stand up today. The deeper you go the harder it is to see how to keep state the same locally and on the server without forcing a refresh. Not sure the best way to do that though since I think we've been dealing with it in a lot of different ways across the project.

Hashiphoto committed:
Refactor accept/reject into separate files [PAR-78]

Hashiphoto committed:
Get branch up to date [PAR-78]

Hashiphoto commented:
Addressed review comments. I moved Accept and Reject buttons into separate files, set the state of the button to "requested" only after the fetch completes, and implemented various other suggestions as well.

lemniscate8 reviewed (approved):
Looks good!
